### PR TITLE
Search API and Alto annotations to S3

### PR DESCRIPTION
**Pull request by tomcrane**
This ports over significant functionality from dds-ecosystem, but now emits annotations to S3 as W3C annotations.

This means, for every manifest:
- a single S3 resource with all that manifest's annotations (from every canvas, concatenated to a single page)
- a single S3 resource with all that manifest's images, identified from OCR. This is where ALTO identifies a figure, image, or composed block (e.g., a table of data).
- and for each canvas in the manifest, an S3 resource with just that canvas's annotations.

In order for this to work locally, I added a Helpers class to Dds.Server.
This rewrites fully qualified domains in the IIIF JSON response to your localhost address, so you can navigate from manifests to annotation lists.

### CODE REVIEW

**Comment by tomcrane** [comment]
This is my question to mull over for next few days

https://github.com/wellcomecollection/iiif-builder/pull/77/files#diff-33d07882707ba922e27d511e3bbe60dc51a16a7c2ca3a970165d55d35dca1099R33

It would be relatively simple to include Open Annotation classes in IIIF.LegacyInclusions, and build the v2/ resources at the same time we build the v3 resources (for annotations).

Is that a good idea? We'd end up with twice as much stuff in S3. ~60m annotation files?

These are very infrequently accessed, unless being harvested.

**Comment by tomcrane** [comment]
Also included in this monster PR is an implementation of the IIIF Search API v1.

This is available on two paths: /search/v0 and /search/v1
The former matches the exact behaviour of the current DDS.
The latter is the correct behaviour for `hits`.

File: src/Wellcome.Dds/Wellcome.Dds/WordsAndPictures/Text.cs [hunk 1]
```diff
@@ -171,7 +171,7 @@ private List<ResultRect> GetRectangles(string s, IList<Word> results, Dictionary
 
         private void AddContext(List<ResultRect> coalescedResults)
         {
-            const int maxResultsWithContext = 100;
+            const int maxResultsWithContext = 200;
```
Comment (tomcrane) [inline src/Wellcome.Dds/Wellcome.Dds/WordsAndPictures/Text.cs:174]: Put this in config

File: src/Wellcome.Dds/Wellcome.Dds.Server/Startup.cs [hunk 2]
```diff
@@ -128,9 +140,14 @@ public void ConfigureServices(IServiceCollection services)
...
             services.AddScoped<IWorkStorageFactory, ArchiveStorageServiceWorkStorageFactory>()
+                .AddScoped<Synchroniser>()
+                .AddScoped<IDds, Repositories.Dds>()
                 .AddScoped<StorageServiceClient>()
+                .AddScoped<IIIIFBuilder, IIIFBuilder>()
                 .AddScoped<IMetsRepository, MetsRepository>()
-                .AddScoped<IDashboardRepository, DashboardRepository>();
+                .AddScoped<IDashboardRepository, DashboardRepository>()
+                .AddScoped<ISearchTextProvider, AltoSearchTextProvider>()
+                .AddScoped<CachingAltoSearchTextProvider>(); // mmm
```
Comment (tomcrane) [inline src/Wellcome.Dds/Wellcome.Dds.Server/Startup.cs:150]: Make a better comment

**Review by donaldgray — APPROVED** [review]

**Pull request merged**